index.rst

```diff
@@ -1,51 +1,1166 @@
+
+.. code-block:: c++
+
+    #include "numpy/arrayobject.h"
+    #include "ndarray/pybind11.h"
+    #include "ndarray/converter.h"
+
+It is also necessary to check that numpy has been installed and setup (otherwise unexpected segfaults will occcur), so in the module definition we add
+
+.. code-block:: c++
+
+    if (_import_array() < 0) {
+            PyErr_SetString(PyExc_ImportError, "numpy.core.multiarray failed to import");
+            return nullptr;
+        }
+
+.. _function_kwargs:
+
+Wrapping Functions with Default Arguments
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The final method remaining to wrap in ``interpolate.h`` is ``makeInterpolate``, which creates an ``Interpolate`` object from the virtual class.
+
+This is an overloaded function, so we define it in the usual way but add ``py::arg("paremeter")`` for *all* of the arguments of the function (not just the ones that we need to give default values).
+In this case
+
+.. code-block:: c++
+
+    mod.def("makeInterpolate", 
+                       (PTR(Interpolate) (*)(std::vector<double> const &,
+                                             std::vector<double> const &,
+                                             Interpolate::Style const)) makeInterpolate,
+                       py::arg("x"), py::arg("y"), py::arg("style")=Interpolate::AKIMA_SPLINE);
+    mod.def("makeInterpolate", 
+                       (PTR(Interpolate) (*)(ndarray::Array<double const, 1> const &,
+                                             ndarray::Array<double const, 1> const &y,
+                                             Interpolate::Style const)) makeInterpolate,
+                       py::arg("x"), py::arg("y"), py::arg("style")=Interpolate::AKIMA_SPLINE);
+
+This can be slightly simplified by adding ``using namespace pybind11::literals;`` to the top of ``interpolate.cc``, which allows us to replace ``py:arg("parameter")`` with ``"parameter"_a``.
 
 .. note::
 
-   **This technote is not yet published.**
+    If pybind11 returns an error during wrapping that the number of arguments does not match, check that you have wrapped all of the arguments with the correct types. Also make sure that you are defining the function in the correct place (ie. is it defined in the module or inside of a class).
+
+.. _wrapped_interpolate:
+
+Wrapped interpolate.cc
+^^^^^^^^^^^^^^^^^^^^^^
+
+When finished ``interpolate.cc`` should look like:
+
+.. code-block:: c++
+
+    #include <pybind11/pybind11.h>
+    //#include <pybind11/operators.h>
+    #include <pybind11/stl.h>
+
+    #include "numpy/arrayobject.h"
+    #include "ndarray/pybind11.h"
+    #include "ndarray/converter.h"
+
+    #include "lsst/afw/math/interpolate.h"
+
+    namespace py = pybind11;
+    using namespace pybind11::literals;
+
+    using namespace lsst::afw::math;
+
+    PYBIND11_DECLARE_HOLDER_TYPE(MyType, std::shared_ptr<MyType>);
```

Comment:
When we first started wrapping you always needed this macro to use shared pointers. Since then Pim or someone else updated pybind11 and it is no longer necessary. This should be removed here.